File: ghidra/scripts/ExtractAll.java
// Ghidra headless script: Extract functions, strings, and imports to JSON
// Usage: analyzeHeadless <project> <name> -import <binary> -postScript ExtractAll.java <output.json>
// @category GenAI-Decompiler

import com.google.gson.JsonArray;
import com.google.gson.JsonObject;
import ghidra.app.script.GhidraScript;
import ghidra.program.model.listing.Function;
import ghidra.program.model.listing.FunctionIterator;
import ghidra.program.model.data.StringDataInstance;
import ghidra.program.model.listing.Data;
import ghidra.program.model.listing.DataIterator;
import ghidra.program.model.symbol.ExternalManager;
import ghidra.program.model.symbol.ExternalLocationIterator;
import ghidra.program.model.symbol.ExternalLocation;

import java.io.FileWriter;
import java.io.IOException;

public class ExtractAll extends GhidraScript {

    @Override
    public void run() throws Exception {
        String[] args = getScriptArgs();
        if (args.length < 1) {
            printerr("Usage: ExtractAll.java <output_path>");
            return;
        }

        String outputPath = args[0];
        JsonObject root = new JsonObject();

        try {
            // Binary metadata
            JsonObject binary = new JsonObject();
            binary.addProperty("name", currentProgram.getName());
            binary.addProperty("arch", currentProgram.getLanguage().getProcessor().toString());
            binary.addProperty("format", currentProgram.getExecutableFormat());
            root.add("binary", binary);

            // Functions
            JsonArray functions = new JsonArray();
            FunctionIterator funcIter = currentProgram.getFunctionManager().getFunctions(true);
            while (funcIter.hasNext() && !monitor.isCancelled()) {
                Function func = funcIter.next();
                JsonObject fn = new JsonObject();
                fn.addProperty("name", func.getName());
                fn.addProperty("address", func.getEntryPoint().toString());
                fn.addProperty("size", func.getBody().getNumAddresses());
                functions.add(fn);
            }
            root.add("functions", functions);

            // Strings
            JsonArray strings = new JsonArray();
            DataIterator dataIter = currentProgram.getListing().getDefinedData(true);
            while (dataIter.hasNext() && !monitor.isCancelled()) {
                Data data = dataIter.next();
                if (data.hasStringValue()) {
                    JsonObject str = new JsonObject();
                    str.addProperty("address", data.getAddress().toString());
                    str.addProperty("value", data.getDefaultValueRepresentation());
                    strings.add(str);
                }
            }
            root.add("strings", strings);

            // Imports
            JsonArray imports = new JsonArray();
            ExternalManager extMgr = currentProgram.getExternalManager();
            for (String libName : extMgr.getExternalLibraryNames()) {
                ExternalLocationIterator extIter = extMgr.getExternalLocations(libName);
                while (extIter.hasNext()) {
                    ExternalLocation extLoc = extIter.next();
                    JsonObject imp = new JsonObject();
                    imp.addProperty("name", extLoc.getLabel());
                    imp.addProperty("library", libName);
                    if (extLoc.getAddress() != null) {
                        imp.addProperty("address", extLoc.getAddress().toString());
                    }
                    imports.add(imp);
                }
            }
            root.add("imports", imports);

            root.addProperty("status", "success");
        } catch (Exception e) {
            root.addProperty("status", "error");
            root.addProperty("error", e.getMessage());
        }

        // Write output
        try (FileWriter writer = new FileWriter(outputPath)) {
            writer.write(root.toString());
        }

        println("ExtractAll: output written to " + outputPath);
    }
}

File: ghidra/scripts/ExtractAll.py
# Ghidra headless script: Extract functions, strings, and imports to JSON
# Usage: analyzeHeadless <project> <name> -import <binary> -postScript ExtractAll.py <output.json>
# @category GenAI-Decompiler
# @runtime Jython

import json

def run():
    args = getScriptArgs()
    if len(args) < 1:
        printerr("Usage: ExtractAll.py <output_path>")
        return

    output_path = args[0]
    root = {}

    try:
        # Binary metadata
        root["binary"] = {
            "name": currentProgram.getName(),
            "arch": str(currentProgram.getLanguage().getProcessor()),
            "format": currentProgram.getExecutableFormat(),
        }

        # Functions
        functions = []
        func_iter = currentProgram.getFunctionManager().getFunctions(True)
        while func_iter.hasNext() and not monitor.isCancelled():
            func = func_iter.next()
            functions.append({
                "name": func.getName(),
                "address": str(func.getEntryPoint()),
                "size": int(func.getBody().getNumAddresses()),
            })
        root["functions"] = functions

        # Strings
        strings = []
        data_iter = currentProgram.getListing().getDefinedData(True)
        while data_iter.hasNext() and not monitor.isCancelled():
            data = data_iter.next()
            if data.hasStringValue():
                strings.append({
                    "address": str(data.getAddress()),
                    "value": data.getDefaultValueRepresentation(),
                })
        root["strings"] = strings

        # Imports
        imports = []
        ext_mgr = currentProgram.getExternalManager()
        for lib_name in ext_mgr.getExternalLibraryNames():
            ext_iter = ext_mgr.getExternalLocations(lib_name)
            while ext_iter.hasNext():
                ext_loc = ext_iter.next()
                imp = {
                    "name": ext_loc.getLabel(),
                    "library": lib_name,
                }
                if ext_loc.getAddress() is not None:
                    imp["address"] = str(ext_loc.getAddress())
                imports.append(imp)
        root["imports"] = imports

        root["status"] = "success"
    except Exception as e:
        root["status"] = "error"
        root["error"] = str(e)

    # Write output
    with open(output_path, "w") as f:
        f.write(json.dumps(root))

    println("ExtractAll: output written to " + output_path)

run()
